File: app/src/main/java/com/channa/mobiledatausageapp/view/MobileDataUsageActivity.java
package com.channa.mobiledatausageapp.view;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import com.channa.mobiledatausageapp.MyApplication;
import com.channa.mobiledatausageapp.R;
import com.channa.mobiledatausageapp.adapter.MobileDataAdapter;
import com.channa.mobiledatausageapp.utility.Utils;
import com.channa.mobiledatausageapp.viewmodel.MobileDataUsageViewModel;

import javax.inject.Inject;

import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;


public class MobileDataUsageActivity extends BaseActivity {

    private static final String TAG = "MobileDataUsageActivity";

    @BindView(R.id.rvMobileDataUsage)
    RecyclerView recyclerViewMobileDataUsage;
    @Inject
    ViewModelProvider.Factory viewModelFactory;
    MobileDataUsageViewModel mobileDataUsageViewModel;
    private MobileDataAdapter mobileDataAdapter;
    private RecyclerView.LayoutManager layoutManager;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mobile_data_usage);

        ((MyApplication) getApplication()).getApplicationComponent().inject(this);

        initViews();
    }

    @Override
    protected void initViews() {
        super.initViews();
        initMobileDataUsageRecyclerView(this);
        getMobileDataUsage();

        if (!Utils.checkInternetConnection(this)) {
            showErrorActionSnackBar("No Internet Connection");
        }
    }

    private void initMobileDataUsageRecyclerView(Context context) {
        recyclerViewMobileDataUsage.setHasFixedSize(true);
        layoutManager = new LinearLayoutManager(this);
        recyclerViewMobileDataUsage.setLayoutManager(layoutManager);
        mobileDataAdapter = new MobileDataAdapter(context);
        recyclerViewMobileDataUsage.setAdapter(mobileDataAdapter);

        mobileDataUsageViewModel = ViewModelProviders.of(this, viewModelFactory).get(MobileDataUsageViewModel.class);
    }

    private void getMobileDataUsage() {
        mobileDataUsageViewModel.getYearlyMobileDataUsage().observe(this, yearListWrapper -> {
            if (null != yearListWrapper.getYearList()) {
                mobileDataAdapter.setYearList(yearListWrapper.getYearList());
                mobileDataAdapter.notifyDataSetChanged();
            } else {
                showErrorActionSnackBar(yearListWrapper.getError());
                Log.e(TAG, "getMobileDataUsage: " + yearListWrapper.getError());
            }
        });
    }
}

File: app/src/main/java/com/zak/datauses/views/MobileDataUsageActivity.kt
package com.zak.datauses.views

import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import com.zak.datauses.MyApplication
import com.zak.datauses.R
import com.zak.datauses.adapters.MobileDataAdapter
import com.zak.datauses.data.model.YearListWrapper
import com.zak.datauses.utilities.Utils.checkInternetConnection
import com.zak.datauses.viewmodel.MobileDataUsageViewModel
import javax.inject.Inject

class MobileDataUsageActivity : BaseActivity() {
    @BindView(R.id.rvMobileDataUsage)
    var recyclerViewMobileDataUsage: RecyclerView? = null

    @JvmField
    @Inject
    var viewModelFactory: ViewModelProvider.Factory? = null
    var mobileDataUsageViewModel: MobileDataUsageViewModel? = null
    private var mobileDataAdapter: MobileDataAdapter? = null
    private var layoutManager: RecyclerView.LayoutManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_data_consumption)
        (application as MyApplication).getApplicationComponent()!!.inject(this)
        recyclerViewMobileDataUsage=findViewById(R.id.rvMobileDataUsage)
        initViews()
    }

    override fun initViews() {
        super.initViews()
        initMobileDataUsageRecyclerView(this)
        mobileDataUsage
        if (!checkInternetConnection(this)) {
            showErrorActionSnackBar("No Internet Connection")
        }
    }

    private fun initMobileDataUsageRecyclerView(context: Context) {
        recyclerViewMobileDataUsage!!.setHasFixedSize(true)
        layoutManager = LinearLayoutManager(this)
        recyclerViewMobileDataUsage!!.layoutManager = layoutManager
        mobileDataAdapter = MobileDataAdapter(context)
        recyclerViewMobileDataUsage!!.adapter = mobileDataAdapter
        mobileDataUsageViewModel = ViewModelProviders.of(this, viewModelFactory).get(MobileDataUsageViewModel::class.java)
    }

    private val mobileDataUsage: Unit
        private get() {
            mobileDataUsageViewModel!!.yearlyMobileDataUsage.observe(this, Observer<YearListWrapper> { yearListWrapper: YearListWrapper ->
                if (null != yearListWrapper.getYearList()) {
                    mobileDataAdapter!!.setYearList(yearListWrapper.getYearList())
                    mobileDataAdapter!!.notifyDataSetChanged()
                } else {
                    showErrorActionSnackBar(yearListWrapper.getError())
                    Log.e(TAG, "getMobileDataUsage: " + yearListWrapper.getError())
                }
            })
        }

    companion object {
        private const val TAG = "MobileDataUsageActivity"
    }
}